storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +257,182 @@ func (s *commitLogSource) Read(
+	namespace ts.ID,
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	series commitlog.Series,
+	dp ts.Datapoint,
+) bool {
+	// Check if the series belongs to current namespace being bootstrapped
+	if !namespace.Equal(series.Namespace) {
+		return false
+	}
+
+	// Check if the shard is one of the shards we're trying to bootstrap
+	ranges, ok := shardsTimeRanges[series.Shard]
+	if !ok {
+		return false
+	}
+
+	// Check if the block corresponds to the time-range that we're trying to bootstrap
+	blockStart := dp.Timestamp.Truncate(blockSize)
+	blockEnd := blockStart.Add(blockSize)
+	blockRange := xtime.Range{
+		Start: blockStart,
+		End:   blockEnd,
+	}
+
+	return ranges.Overlaps(blockRange)
+}
+
+func (s *commitLogSource) printM3TSZEncodingOutcome(workerErrs []int, iter commitlog.Iterator) {
+	errSum := 0
+	for _, numErrs := range workerErrs {
+		errSum += numErrs
+	}
+	if errSum > 0 {
+		s.log.Errorf("error bootstrapping from commit log: %d block encode errors", errSum)
 	}
 	if err := iter.Err(); err != nil {
 		s.log.Errorf("error reading commit log: %v", err)
 	}
+}
 
-	errs = 0
-	emptyErrs := 0
+func (s *commitLogSource) mergeShards(
+	numShards int,
+	bopts result.Options,
+	blopts block.Options,
+	encoderPool encoding.EncoderPool,
+	unmerged []map[ts.Hash]encodersByTime,
+) result.BootstrapResult {
+	shardErrs := make([]int, numShards)
+	shardEmptyErrs := make([]int, numShards)
 	bootstrapResult := result.NewBootstrapResult()
 	blocksPool := bopts.DatabaseBlockOptions().DatabaseBlockPool()
 	multiReaderIteratorPool := blopts.MultiReaderIteratorPool()
+	// Controls how many shards can be merged in parallel
+	mergeSemaphore := make(chan bool, s.opts.MergeShardsConcurrency())
+	bootstrapResultLock := sync.Mutex{}
+	wg := &sync.WaitGroup{}
+
 	for shard, unmergedShard := range unmerged {
-		shardResult := result.NewShardResult(len(unmergedShard), s.opts.ResultOptions())
-		for _, unmergedBlocks := range unmergedShard {
-			blocks := block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
-			for start, unmergedBlock := range unmergedBlocks.encoders {
-				block := blocksPool.Get()
-				if len(unmergedBlock) == 0 {
-					emptyErrs++
-					continue
-				} else if len(unmergedBlock) == 1 {
-					block.Reset(start, unmergedBlock[0].enc.Discard())
-				} else {
-					readers := make([]io.Reader, len(unmergedBlock))
-					for i := range unmergedBlock {
-						stream := unmergedBlock[i].enc.Stream()
-						readers[i] = stream
-						defer stream.Finalize()
-					}
+		mergeSemaphore <- true
+		wg.Add(1)
+		go func(shard int, unmergedShard map[ts.Hash]encodersByTime) {
+			var shardResult result.ShardResult
+			shardResult, shardEmptyErrs[shard], shardErrs[shard] = s.mergeShard(
+				unmergedShard, blocksPool, multiReaderIteratorPool, encoderPool, blopts)
+			if len(shardResult.AllSeries()) > 0 {
+				// Prevent race conditions while updating bootstrapResult from multiple go-routines
```

Comment:
If we separated out a set of workers that each got a certain set of shards similar to the encoder workers we could avoid locking here too.